README.md
<!-- Copyright (c) Meta Platforms, Inc. and affiliates.
All rights reserved.

This source code is licensed under the license found in the
LICENSE file in the root directory of this source tree.
-->

# WhatsApp Business Platform Node.js SDK for the Cloud API, hosted by Meta
<p align="center">
<img src="./website/static/img/wa_logo-216px.svg" width="216" alt="WhatsApp Logo" />
</p>

Welcome to SDK for the [WhatsApp Business Platform](https://business.whatsapp.com/products/business-platform/). This SDK is written for Node.js framework to simplify access to the [Cloud API](https://developers.facebook.com/docs/whatsapp/cloud-api/). The source code itself is written in Typescript with TypeScript declaration files to type-check usage of the WhatsApp Business Platform Node.js SDK in your code, along with hints and code completion in TypeScript compatible IDEs.

[![lint, prettify, spellcheck, test, and build](https://github.com/WhatsApp/WhatsApp-Nodejs-SDK/actions/workflows/nodejs.ci.yml/badge.svg)](https://github.com/WhatsApp/WhatsApp-Nodejs-SDK/blob/main/.github/workflows/nodejs.ci.yml)
[![generate docs](https://github.com/WhatsApp/WhatsApp-Nodejs-SDK/actions/workflows/docusaurus.yml/badge.svg)](https://github.com/WhatsApp/WhatsApp-Nodejs-SDK/blob/main/.github/workflows/docusaurus.yml)

## Getting started
View the [quick start documentation](https://whatsapp.github.io/WhatsApp-Nodejs-SDK/) to learn how to use the SDK and get started.

## Installation
Install the WhatsApp Business Platform SDK using yarn:

```shell
yarn add whatsapp
```

Or npm:

```shell
npm install whatsapp
```

## Configuration
The SDK uses environmental variables for setting all the configuration. For development purposes, you can use a **.env** file at the root of your project. Below are all the possible options for the SDK configuration and only some are required for certain features.

```shell
# The base URL to send all SDK requests to (default graph.facebook.com).
# This variable should not be used unless necessary for development or special routing needs.
WA_BASE_URL=

# Your Meta for Developers app Id.
M4D_APP_ID=

# Your Meta for Developers Business app secret.
M4D_APP_SECRET=

# Your WhatsApp phone number Id (sender).
WA_PHONE_NUMBER_ID=

# Your WhatsApp business account Id.
WA_BUSINESS_ACCOUNT_ID=

# System user access token. Recommended: Do not use a temporary access token.
CLOUD_API_ACCESS_TOKEN=

# Cloud API version number.
CLOUD_API_VERSION=v16.0

# Customize your incoming webhook listener endpoint. Path should be
# https://{host}/{WEBHOOK_ENDPOINT}. A trailing slash is not added by default,
# so the variable should include that if it's required by your API gateway.
WEBHOOK_ENDPOINT=

# A custom verification token string to validate incoming webhook payloads.
# Needs to match webhook configuration.
WEBHOOK_VERIFICATION_TOKEN=

# Override the default app listener port (port 3000).
LISTENER_PORT=

# Turn on global debug logging
DEBUG=

# The number of request retries after waiting (default 30 retries)
MAX_RETRIES_AFTER_WAIT=

# The timeout period in milliseconds for a request to wait for a response (default 20000ms)
REQUEST_TIMEOUT=
```

## Code of Conduct
Meta has adopted a Code of Conduct that we expect project participants to adhere to. Please read the full text so that you can understand what actions will and will not be tolerated.

## Contribute
See the [CONTRIBUTING](CONTRIBUTING.md) file for our development process, how to propose bugfixes and improvements, and how to build and test your changes to the WhatsApp Business Platform Node.js SDK.

## License
The WhatsApp Business Platform Node.js SDK for the Cloud API is Meta Platforms licensed, as found in the LICENSE file.


CHANGELOG

# Change Log
All notable changes to this project are documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/)
and this project adheres to [Semantic Versioning](http://semver.org/).

## [0.0.5-Alpha] - 2023-04-11
### Added
- Two-step verification API.
- Docs file names match pascal/camel case for each of the APIs.
- Docs for missing types.
- Fix for incorrect parameter name for language object.

### Removed
- Dotenv to be a dev only dependency.
- Typos from docs.
- Unnecessary extra nesting for interactive messages requests.

## [0.0.4-Alpha] - 2023-04-05

### Added
- Prettier covers everything, including docs website.
- Update dev dependencies.
- Expand lint coverage.

### Removed
- Support for Node 14.x.
- Build directory from repository.
- Erroneous or duplicate options from TSC config.

## [0.0.3-Alpha] - 2023-04-03

### Added
- Update docusaurus.
- Update styling to use pascal or camel case and other suggestions from https://google.github.io/styleguide/tsguide.html#identifiers to resolve https://github.com/WhatsApp/WhatsApp-Nodejs-SDK/issues/1.
- Fix broken types for reply buttons.
- Reduce number of "any" for variable types.
- Improve docs.

## [0.0.2-Alpha] - 2023-03-29

### Added
- Support transpiled code down to ES2015.
- Add babel fix for ES6 import without needing to specify default.
- Fix typos in docs and update module import description.

## [0.0.1-Alpha] - 2023-03-20

### Added
- Messages endpoints, webhooks listener w/ validation.


jest.config.ts
/**
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 * All rights reserved.
 *
 * This source code is licensed under the license found in the
 * LICENSE file in the root directory of this source tree.
 */

/*
 * For a detailed explanation regarding each configuration property, visit:
 * https://jestjs.io/docs/configuration
 */

import type { Config } from 'jest';

const config: Config = {
	// All imported modules in your tests should be mocked automatically
	// automock: false,

	// Stop running tests after `n` failures
	// bail: 0,

	// The directory where Jest should store its cached dependency information
	// cacheDirectory: "/private/var/folders/s4/94j4h24s0yzc8cbg2_xddt3h0000gn/T/jest_dx",

	// Automatically clear mock calls, instances, contexts and results before every test
	clearMocks: true,

	// Indicates whether the coverage information should be collected while executing the test
	collectCoverage: true,

	// An array of glob patterns indicating a set of files for which coverage information should be collected
	// collectCoverageFrom: undefined,

	// The directory where Jest should output its coverage files
	coverageDirectory: 'coverage',

	// An array of regexp pattern strings used to skip coverage collection
	// coveragePathIgnorePatterns: [
	//   "/node_modules/"
	// ],

	// Indicates which provider should be used to instrument code for coverage
	// coverageProvider: "babel",

	// A list of reporter names that Jest uses when writing coverage reports
	// coverageReporters: [
	//   "json",
	//   "text",
	//   "lcov",
	//   "clover"
	// ],

	// An object that configures minimum threshold enforcement for coverage results
	// coverageThreshold: undefined,

	// A path to a custom dependency extractor
	// dependencyExtractor: undefined,

	// Make calling deprecated APIs throw helpful error messages
	// errorOnDeprecated: false,

	// The default configuration for fake timers
	// fakeTimers: {
	//   "enableGlobally": false
	// },

	// Force coverage collection from ignored files using an array of glob patterns
	// forceCoverageMatch: [],

	// A path to a module which exports an async function that is triggered once before all test suites
	// globalSetup: undefined,

	// A path to a module which exports an async function that is triggered once after all test suites
	// globalTeardown: undefined,

	// A set of global variables that need to be available in all test environments
	globals: {
		sdkConfig: {
			WA_BASE_URL: `example.com`,
			M4D_APP_ID: 'TEST_APP_ID',
			M4D_APP_SECRET: 'TEST_APP_SECRET',
			WA_PHONE_NUMBER_ID: 1234567890,
			WA_BUSINESS_ACCOUNT_ID: 'TEST_WA_BUSINESS_ACCOUNT_ID',
			CLOUD_API_ACCESS_TOKEN: 'TEST_CLOUD_API_ACCESS_TOKEN',
			CLOUD_API_VERSION: 'TEST_v16',
			WEBHOOK_ENDPOINT: 'TEST_WEBHOOK_ENDPOINT',
			WEBHOOK_VERIFICATION_TOKEN: 'TEST_WEBHOOK_VERIFICATION_TOKEN',
			LISTENER_PORT: 3000,
			DEBUG: true,
			MAX_RETRIES_AFTER_WAIT: 30,
			REQUEST_TIMEOUT: 900,
		},
	},

	// The maximum amount of workers used to run your tests. Can be specified as % or a number. E.g. maxWorkers: 10% will use 10% of your CPU amount + 1 as the maximum worker number. maxWorkers: 2 will use a maximum of 2 workers.
	// maxWorkers: "50%",

	// An array of directory names to be searched recursively up from the requiring module's location
	// moduleDirectories: [
	//   "node_modules"
	// ],

	// An array of file extensions your modules use
	moduleFileExtensions: ['ts', 'tsx', 'js', 'jsx', 'json', 'node'],

	// A map from regular expressions to module names or to arrays of module names that allow to stub out resources with a single module
	// moduleNameMapper: {},

	// An array of regexp pattern strings, matched against all module paths before considered 'visible' to the module loader
	// modulePathIgnorePatterns: [],

	// Activates notifications for test results
	// notify: false,

	// An enum that specifies notification mode. Requires { notify: true }
	// notifyMode: "failure-change",

	// A preset that is used as a base for Jest's configuration
	// preset: undefined,

	// Run tests from one or more projects
	// projects: undefined,

	// Use this configuration option to add custom reporters to Jest
	// reporters: undefined,

	// Automatically reset mock state before every test
	// resetMocks: false,

	// Reset the module registry before running each individual test
	// resetModules: false,

	// A path to a custom resolver
	// resolver: undefined,

	// Automatically restore mock state and implementation before every test
	// restoreMocks: false,

	// The root directory that Jest should scan for tests and modules within
	// rootDir: undefined,

	// A list of paths to directories that Jest should use to search for files in
	roots: ['<rootDir>/src'],

	// Allows you to use a custom runner instead of Jest's default test runner
	// runner: "jest-runner",

	// The paths to modules that run some code to configure or set up the testing environment before each test
	// setupFiles: [],

	// A list of paths to modules that run some code to configure or set up the testing framework before each test
	// setupFilesAfterEnv: [],

	// The number of seconds after which a test is considered as slow and reported as such in the results.
	// slowTestThreshold: 5,

	// A list of paths to snapshot serializer modules Jest should use for snapshot testing
	// snapshotSerializers: [],

	// The test environment that will be used for testing
	testEnvironment: 'node',

	// Options that will be passed to the testEnvironment
	// testEnvironmentOptions: {},

	// Adds a location field to test results
	// testLocationInResults: false,

	// The glob patterns Jest uses to detect test files
	// testMatch: [
	//   "**/__tests__/**/*.[jt]s?(x)",
	//   "**/?(*.)+(spec|test).[tj]s?(x)"
	// ],

	// An array of regexp pattern strings that are matched against all test paths, matched tests are skipped
	// testPathIgnorePatterns: [
	//   "/node_modules/"
	// ],

	// The regexp pattern or array of patterns that Jest uses to detect test files
	testRegex: './src/.*\\.(test|spec)?\\.(ts|ts)$',

	// This option allows the use of a custom results processor
	// testResultsProcessor: undefined,

	// This option allows use of a custom test runner
	// testRunner: "jest-circus/runner",

	// A map from regular expressions to paths to transformers
	transform: {
		'^.+\\.ts?$': 'ts-jest',
	},

	// An array of regexp pattern strings that are matched against all source file paths, matched files will skip transformation
	// transformIgnorePatterns: [
	//   "/node_modules/",
	//   "\\.pnp\\.[^\\/]+$"
	// ],

	// An array of regexp pattern strings that are matched against all modules before the module loader will automatically return a mock for them
	// unmockedModulePathPatterns: undefined,

	// Indicates whether each individual test should be reported during the run
	// verbose: undefined,

	// An array of regexp patterns that are matched against all source file paths before re-running tests in watch mode
	// watchPathIgnorePatterns: [],

	// Whether to use watchman for file crawling
	// watchman: true,
};

export default config;
